### PR TITLE
Feature/reward actor

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Reward actor v0 and v2.

### CODE REVIEW

File: core/vm/actor/builtin/v0/reward/reward_actor_state.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                           .velocity = kInitialRewardVelocityEstimate},
+        .this_epoch_baseline_power =
+            initBaselinePower(kBaselineInitialValueV0, kBaselineExponentV0),
+        .epoch = kChainEpochUndefined,
+        .total_mined = {0},
+    };
+    updateToNextEpochWithReward(
+        state, current_realized_power, kBaselineExponentV0);
+    return state;
+  }
```
Comment (wer1st) [inline core/vm/actor/builtin/v0/reward/reward_actor_state.cpp:31]: Could we move it to constructor?
State(const StoragePower &current_realized_power)

File: core/vm/actor/builtin/v2/reward/reward_actor.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  ACTOR_METHOD_IMPL(AwardBlockReward) {
+    OUTCOME_TRY(balance,
+                v0::reward::AwardBlockReward::validateParams(runtime, params));
+    OUTCOME_TRY(miner, runtime.resolveAddress(params.miner));
+    OUTCOME_TRY(state, runtime.getCurrentActorStateCbor<State>());
+    OUTCOME_TRY(reward,
+                v0::reward::AwardBlockReward::calculateReward(
+                    runtime, params, state.this_epoch_reward, balance));
+    const auto [block_reward, total_reward] = reward;
```
Comment (wer1st) [inline core/vm/actor/builtin/v2/reward/reward_actor.cpp:32]:
```suggestion
    const auto& [block_reward, total_reward] = reward;
```

File: core/vm/actor/builtin/v2/reward/reward_actor.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    state.total_storage_power_reward += block_reward;
+    OUTCOME_TRY(runtime.commitState(state));
+
+    VM_ASSERT(total_reward <= balance);
+
+    const TokenAmount penalty = kPenaltyMultiplier * params.penalty;
+    const ApplyRewards::Params reward_params{.reward = total_reward,
+                                             .penalty = penalty};
+    auto expected = penalty.str();
```
Comment (wer1st) [inline core/vm/actor/builtin/v2/reward/reward_actor.cpp:41]: Never used

File: core/vm/exit_code/exit_code.hpp [hunk 4]
```diff
@@ -10,9 +10,9 @@
 
 #include "common/outcome.hpp"
 
-#define VM_ASSERT(condition)                      \
-  if (!(condition)) {                             \
-    return outcome::failure(VMExitCode::kAssert); \
+#define VM_ASSERT(condition)                                       \
+  if (!(condition)) {                                              \
+    return outcome::failure(VMAbortExitCode{VMExitCode::kAssert}); \
```
Comment (wer1st) [inline core/vm/exit_code/exit_code.hpp:15]:
```suggestion
    return outcome::failure(static_cast<VMAbortExitCode>(VMExitCode::kAssert)); \
```

File: test/core/common/math/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#
+# Copyright Soramitsu Co., Ltd. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+addtest(math_test
+    math_test.cpp
+    )
```
Comment (wer1st) [inline test/core/common/math/CMakeLists.txt:8]: minor: add an empty line as the last line

File: test/core/vm/actor/builtin/v0/reward/reward_actor_v0_test.cpp [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    const TokenAmount balance{9};
+    setCurrentBalance(balance);
+
+    const Address winner = Address::makeFromId(1000);
+    const TokenAmount gas_reward{10};
+    auto res = AwardBlockReward::call(runtime, {winner, 0, gas_reward, 1});
+
+    EXPECT_TRUE(res.has_error());
+    auto error = res.error();
+    EXPECT_TRUE(isAbortExitCode(error));
```
Comment (wer1st) [inline test/core/vm/actor/builtin/v0/reward/reward_actor_v0_test.cpp:141]: Redundant check. It is checked below as well.

File: test/core/vm/actor/builtin/v0/reward/reward_actor_v0_test.cpp [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    EXPECT_EQ(VMAbortExitCode{VMExitCode::kErrIllegalState}, error);
```
Comment (wer1st) [inline test/core/vm/actor/builtin/v0/reward/reward_actor_v0_test.cpp:142]:
```suggestion
    EXPECT_EQ(AwardBlockReward::call(runtime, {winner, 0, gas_reward, 1}), VMAbortExitCode{VMExitCode::kErrIllegalState});
```

Comment (wer1st) [reply test/core/vm/actor/builtin/v0/reward/reward_actor_v0_test.cpp:142]: For all tests

**Review by turuslan — APPROVED** [review]

File: core/vm/actor/builtin/v0/reward/reward_actor.cpp [hunk 8]
```diff
@@ -6,102 +6,98 @@
...
+    // Reduce the payable reward by the penalty.
+    const TokenAmount reward_payable = total_reward - penalty;
+    VM_ASSERT(reward_payable + penalty <= balance);
+
+    // if this fails, we can assume the miner is responsible and avoid failing
+    // here.
+    const auto res = runtime.sendM<miner::AddLockedFund>(
+        miner, reward_payable, reward_payable);
+    if (res.has_error()) {
+      OUTCOME_TRY(runtime.sendFunds(kBurntFundsActorAddress, reward_payable));
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/v0/reward/reward_actor.cpp:75]: Yep, just burn them if AddLockedFund has failed.

[the remark this answers is not in the document]
Comment (wer1st) [reply core/vm/actor/builtin/v0/reward/reward_actor.cpp:75]: Ok

File: core/vm/actor/builtin/v0/reward/reward_actor_state.cpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/v0/reward/reward_actor_state.cpp:31]: Yes, but default constructor is required. Is it better?

File: core/vm/actor/builtin/v0/reward/reward_actor_state.hpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+   * Q.128 formatted number such that f(epoch) = baseExponent^epoch grows 100%
+   * in one year of epochs
+   * Calculation here:
+   * https://www.wolframalpha.com/input/?i=IntegerPart%5BExp%5BLog%5B1%2B100%25%5D%2F%28%28365+days%29%2F%2830+seconds%29%29%5D*2%5E128%5D
+   */
+  static const BigInt kBaselineExponentV3{
+      "340282591298641078465964189926313473653"};
+
+  struct State {
+    State();
```
Comment (wer1st) [inline core/vm/actor/builtin/v0/reward/reward_actor_state.hpp:59]:
```suggestion
    State() = default;
```

Comment (wer1st) [reply core/vm/actor/builtin/v0/reward/reward_actor_state.hpp:59]: The same for v2

File: core/vm/actor/builtin/v0/reward/reward_actor_state.cpp (hunk 1 above)
Comment (wer1st) [reply core/vm/actor/builtin/v0/reward/reward_actor_state.cpp]: Yes, it is ok

**Review by wer1st — APPROVED** [review]

**Pull request merged**